### PR TITLE
Preserve varnames info, display in mermaid, fix #327

### LINKED ISSUE #327: associate graph nodes with users's hydroflow variables, plot appropriately
It would be helpful for the mermaid/dot graphs to show the users' hydroflow variable names.  Perhaps as (nested?) subgraphs.

To do this we have to propagate the hydroflow variable names into `FlatGraph`, `PartitionedGraph` and `SerdeGraph`.

### PR DESCRIPTION
**Pull request by MingweiSamuel**
Still need to include info in `write_dot`

Also fixes some awkward redundancies in the `PartitionedGraph::to_serde_graph` code

fix #327

### CODE REVIEW

File: hydroflow_lang/src/graph/serde_graph.rs [hunk 1]
```diff
@@ -198,6 +227,24 @@ impl SerdeGraph {
                 }
             }
 
+            // write out any variable names
+            for (varname, varname_node_ids) in self.varname_nodes.iter() {
+                // TODO(mingwei): this is awkward, inefficient runtime
+                let mut local_named_nodes = varname_node_ids
+                    .iter()
+                    .filter(|node_id| node_ids.contains(node_id))
+                    .peekable();
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/serde_graph.rs:236]: Not particularly efficient/elegant

**Comment by MingweiSamuel** [comment]
```rust
        // inputs: the origin vertex (node 0) and stream of input edges
        origin = source_iter(vec![0]);
        stream_of_edges = source_stream(edges_recv);
        reached_vertices = merge();
        origin -> [0]reached_vertices;

        // the join
        my_join_tee = join() -> flat_map(|(src, ((), dst))| [src, dst]) -> tee();
        reached_vertices -> map(|v| (v, ())) -> [0]my_join_tee;
        stream_of_edges -> [1]my_join_tee;

        // the loop and the output
        my_join_tee[0] -> [1]reached_vertices;
        my_join_tee[1] -> unique() -> for_each(|x| println!("Reached: {}", x));
```
updated mermaid output:
```mermaid
%%{init:{'theme':'base','themeVariables':{'clusterBkg':'#ddd','clusterBorder':'#888'}}}%%
flowchart TD
classDef pullClass fill:#02f,color:#fff,stroke:#000
classDef pushClass fill:#ff0,stroke:#000
linkStyle default stroke:#aaa,stroke-width:4px,color:red,font-size:1.5em;
subgraph sg_1v1 ["sg_1v1 stratum 0"]
    1v1[\"(1v1) <tt>source_iter(vec! [0])</tt>"/]:::pullClass
    2v1[\"(2v1) <tt>source_stream(edges_recv)</tt>"/]:::pullClass
    3v1[\"(3v1) <tt>merge()</tt>"/]:::pullClass
    7v1[\"(7v1) <tt>map(| v | (v, ()))</tt>"/]:::pullClass
    4v1[\"(4v1) <tt>join()</tt>"/]:::pullClass
    5v1[/"(5v1) <tt>flat_map(| (src, ((), dst)) | [src, dst])</tt>"\]:::pushClass
    6v1[/"(6v1) <tt>tee()</tt>"\]:::pushClass
    10v1["(10v1) <tt>handoff</tt>"]:::otherClass
    10v1--1--->3v1
    1v1--0--->3v1
    2v1--1--->4v1
    3v1--->7v1
    7v1--0--->4v1
    4v1--->5v1
    5v1--->6v1
    6v1--0--->10v1
    subgraph sg_1v1_var_my_join_tee ["var <tt>my_join_tee</tt>"]
        4v1
        5v1
        6v1
    end
    subgraph sg_1v1_var_origin ["var <tt>origin</tt>"]
        1v1
    end
    subgraph sg_1v1_var_reached_vertices ["var <tt>reached_vertices</tt>"]
        3v1
    end
    subgraph sg_1v1_var_stream_of_edges ["var <tt>stream_of_edges</tt>"]
        2v1
    end
end
subgraph sg_2v1 ["sg_2v1 stratum 1"]
    8v1[/"(8v1) <tt>unique()</tt>"\]:::pushClass
    9v1[/"(9v1) <tt>for_each(| x | println! (&quot;Reached: {}&quot;, x))</tt>"\]:::pushClass
    8v1--->9v1
end
6v1--1--->11v1
11v1["(11v1) <tt>handoff</tt>"]:::otherClass
11v1===o8v1
```

note that if a variable name spans across subgraphs it will appear as separate boxes, one in each subgraph

File: hydroflow_lang/src/graph/flat_graph.rs [hunk 2]
```diff
@@ -34,8 +35,10 @@ pub struct FlatGraph {
     /// Spanned error/warning/etc diagnostics to emit.
     pub(crate) diagnostics: Vec<Diagnostic>,
 
+    /// What variable name each graph node belongs to (if any).
+    pub(crate) node_varnames: SparseSecondaryMap<GraphNodeId, Rc<Ident>>,
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/flat_graph.rs:39]: I'm curious when it's worth using an Rc ref and when to copy. I assume `Ident` is small?

File: hydroflow_lang/src/graph/serde_graph.rs [hunk 3]
```diff
@@ -39,9 +44,11 @@ impl SerdeGraph {
 
         fn write_mermaid_prelude(write: &mut impl std::fmt::Write) -> std::fmt::Result {
             // intro
+            #[allow(clippy::write_literal)]
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/serde_graph.rs:47]: i don't understand why it's better to write this way?

File: hydroflow_lang/src/graph/serde_graph.rs [hunk 4]
```diff
@@ -135,11 +142,11 @@ impl SerdeGraph {
         ) -> std::fmt::Result {
             writeln!(
                 write,
-                "{:t$}subgraph \"sg_{:?} stratum {:?}\"",
+                "{:t$}subgraph sg_{sg:?} [\"sg_{sg:?} stratum {:?}\"]",
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/serde_graph.rs:145]: did I have those printing backward?

**Review by jhellerstein — APPROVED** [review]
LGTM

File: hydroflow_lang/src/graph/flat_graph.rs (hunk 2 above)
Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/flat_graph.rs:39]: Hmm, yeah probably better and definitely simpler to just `Clone`

File: hydroflow_lang/src/graph/serde_graph.rs (hunk 3 above)
Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/serde_graph.rs]: changed it back, not a big difference, just whether the `{{` are escaped or not

File: hydroflow_lang/src/graph/serde_graph.rs (hunk 4 above)
Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/serde_graph.rs:143]: No, just providing a separate shorter ID for the subgraphs (`sg_#`) vs their label (`sg_# stratum #`)

**Pull request merged**